### PR TITLE
Add tracing decorator

### PR DESCRIPTION
**Pull request by shezadkhan137**

### CODE REVIEW

File: decorators/tracing/receiver_test.go [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+	}
+
+	b64 := base64.StdEncoding.EncodeToString(propagation.Binary(sc))
+	return sc, b64
+}
+
+// Tests that when a Receiver is wrapped by TracingReceiver, and tracecontext
+// is present, a span is started and set in the receive context with the correct
+// parent context
+func TestDecoder_SuccessfullyDecodesSpanWhenTraceContextIsPresent(t *testing.T) {
```
Comment (Xopherus) [inline decorators/tracing/receiver_test.go]: ❤️ these tests. I'm actually surprised golangci-lint didn't complain about duplicated code - some of the assertions look familiar.

File: decorators/tracing/topic.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+type tracingWriter struct {
+	Next msg.MessageWriter
+
+	buf     bytes.Buffer
+	closed  bool
+	mux     sync.Mutex
+	ctx     context.Context
+	onClose func()
```
Comment (Xopherus) [inline decorators/tracing/topic.go]: I think we should call out in a godoc what `onClose` is used for. The rest of the fields are shared by others, so it's less critical they are called out in godoc.

File: decorators/tracing/receiver.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	}
+
+	for _, opt := range opts {
+		opt(options)
+	}
+
+	return msg.ReceiverFunc(func(ctx context.Context, m *msg.Message) error {
+		ctx, span := withContext(ctx, m, options)
+		defer span.End()
+		select {
```
[the remark this answers is not in the document]
Comment (shezadkhan137) [reply decorators/tracing/receiver.go]: Cool. Yeh I saw it done in the other decorator :-)

File: decorators/tracing/receiver.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			return next.Receive(ctx, m)
+		}
+	})
+}
+
+func withContext(ctx context.Context, m *msg.Message, options *Options) (context.Context, *trace.Span) {
+	attrs := m.Attributes
```
[the remark this answers is not in the document]
Comment (shezadkhan137) [reply decorators/tracing/receiver.go]: No, I was setting it via the map before. Will change.

File: decorators/tracing/receiver.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			return next.Receive(ctx, m)
+		}
+	})
+}
+
+func withContext(ctx context.Context, m *msg.Message, options *Options) (context.Context, *trace.Span) {
```
[the remark this answers is not in the document]
Comment (shezadkhan137) [reply decorators/tracing/receiver.go]: 👍

File: decorators/tracing/receiver_test.go (hunk 1 above)
Comment (shezadkhan137) [reply decorators/tracing/receiver_test.go]: Haha, shamelessly copied from the b64 decorator

File: decorators/tracing/topic.go (hunk 2 above)
Comment (shezadkhan137) [reply decorators/tracing/topic.go]: 👍

File: decorators/tracing/doc.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+// into the context.Context object which is passed into the receiver that you
+// provide handle messages. Again if to trace is present a trace is started and
+// set in the context.
+//
+// Examples
+//
+// Using the tracing.Topic:
+//
+//	func ExampleTopic() {
+//      // make a concrete topic eg SNS
```
Comment (Xopherus) [inline decorators/tracing/doc.go]: Indentation is a little off here btw

File: decorators/tracing/receiver.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+	}
+}
+
+func WithStartOption(so trace.StartOptions) Option {
+	return func(o *Options) {
+		o.StartOptions = so
+	}
+}
+
+// Wraps another msg.Receiver, populating the context
```
Comment (Xopherus) [inline decorators/tracing/receiver.go]: Godoc comments for public methods/types should start with the type. So in this case I would simply prefix this sentence like so:

```
// Receiver wraps another msg.Receiver, populating ... 
```

**Pull request merged**